File: sdk/java/src/main/java/com/xxxxteam/tempmail/providers/Altmails.java
package com.xxxxteam.tempmail.providers;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.xxxxteam.tempmail.Email;
import com.xxxxteam.tempmail.EmailInfo;
import com.xxxxteam.tempmail.HttpResult;
import com.xxxxteam.tempmail.HttpClient;
import com.xxxxteam.tempmail.Json;
import com.xxxxteam.tempmail.Normalizer;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Altmails 渠道 — https://tempmail.altmails.com
 *
 * <p>GET / 获取 session 和 CSRF → GET /random-email-address 获取随机邮箱
 * → POST /fetch-emails/{email} 获取邮件列表 → GET /view/{id} 获取正文。</p>
 */
public final class Altmails {

    private static final String BASE_URL = "https://tempmail.altmails.com";
    private static final Pattern CSRF_RE = Pattern.compile("['\"]_token['\"]\\s*:\\s*['\"]([^'\"]+)['\"]");

    private Altmails() {
    }

    private static Map<String, String> browserHeaders(String extra) {
        Map<String, String> h = new LinkedHashMap<>();
        h.put("Accept", "text/html,application/xhtml+xml");
        h.put("User-Agent", "Mozilla/5.0");
        return h;
    }

    private static String cookieStr(HttpResult resp) {
        StringBuilder sb = new StringBuilder();
        for (String sc : resp.getSetCookies()) {
            String seg = sc.split(";")[0].trim();
            if (sb.length() > 0) sb.append("; ");
            sb.append(seg);
        }
        return sb.toString();
    }

    private static String mergeCookieStr(String prev, HttpResult resp) {
        Map<String, String> jar = new LinkedHashMap<>();
        if (prev != null) {
            for (String part : prev.split(";")) {
                String p = part.trim();
                int eq = p.indexOf('=');
                if (eq > 0) jar.put(p.substring(0, eq).trim(), p.substring(eq + 1).trim());
            }
        }
        for (String sc : resp.getSetCookies()) {
            String seg = sc.split(";")[0].trim();
            int eq = seg.indexOf('=');
            if (eq > 0) jar.put(seg.substring(0, eq), seg.substring(eq + 1));
        }
        StringBuilder sb = new StringBuilder();
        for (var e : jar.entrySet()) {
            if (sb.length() > 0) sb.append("; ");
            sb.append(e.getKey()).append("=").append(e.getValue());
        }
        return sb.toString();
    }

    /**
     * 创建 altmails 临时邮箱。
     *
     * @return 邮箱信息
     */
    public static EmailInfo generate() {
        // 获取首页，建立 session 并提取 CSRF
        HttpResult r1 = HttpClient.get(BASE_URL + "/", browserHeaders(null));
        r1.ensureSuccess();
        String cookies = cookieStr(r1);

        Matcher m = CSRF_RE.matcher(r1.getBody());
        if (!m.find()) throw new RuntimeException("altmails: 无法从首页提取 CSRF token");
        String csrf = m.group(1);

        // GET /random-email-address 获取随机邮箱地址
        Map<String, String> h2 = new LinkedHashMap<>();
        h2.put("Accept", "text/html");
        h2.put("User-Agent", "Mozilla/5.0");
        h2.put("Cookie", cookies);
        h2.put("Referer", BASE_URL + "/");
        HttpResult r2 = HttpClient.get(BASE_URL + "/random-email-address", h2);
        r2.ensureSuccess();

        cookies = mergeCookieStr(cookies, r2);
        String mailbox = r2.getBody().trim();
        if (mailbox.isEmpty() || !mailbox.contains("@")) {
            throw new RuntimeException("altmails: 返回的邮箱地址无效");
        }

        String token = Json.serialize(Map.of("csrf", csrf, "cookies", cookies));
        return new EmailInfo("altmails", mailbox, token, null, null);
    }

    /**
     * 获取邮件列表。
     *
     * @param email 邮箱地址
     * @param token JSON 编码的 {csrf, cookies}
     * @return 邮件列表
     */
    public static List<Email> getEmails(String email, String token) {
        if (token == null || token.isBlank()) throw new RuntimeException("altmails: token 为空");
        if (email == null || email.isBlank()) throw new RuntimeException("altmails: 邮箱地址为空");

        JsonObject tokObj = Json.parseObject(token);
        if (tokObj == null) throw new RuntimeException("altmails: token 格式无效");
        String csrf = Json.str(tokObj, "csrf");
        String cookies = Json.str(tokObj, "cookies");

        // POST /fetch-emails/{email} 获取邮件列表
        Map<String, String> h = new LinkedHashMap<>();
        h.put("Accept", "application/json");
        h.put("X-Requested-With", "XMLHttpRequest");
        h.put("Content-Type", "application/x-www-form-urlencoded");
        h.put("Cookie", cookies);
        h.put("Referer", BASE_URL + "/");
        h.put("User-Agent", "Mozilla/5.0");

        HttpResult resp = HttpClient.post(BASE_URL + "/fetch-emails/" + email,
                "_token=" + ProviderUtil.urlEncode(csrf),
                "application/x-www-form-urlencoded", h);
        resp.ensureSuccess();

        JsonElement parsed = Json.parse(resp.getBody());
        if (parsed == null || !parsed.isJsonArray()) return new ArrayList<>();

        List<Email> out = new ArrayList<>();
        for (JsonElement item : parsed.getAsJsonArray()) {
            if (!item.isJsonObject()) continue;
            JsonObject msg = item.getAsJsonObject();
            String mailId = Json.str(msg, "id");

            // 尝试获取 HTML 正文
            String htmlBody = "";
            if (!mailId.isEmpty()) {
                try {
                    Map<String, String> vh = new LinkedHashMap<>();
                    vh.put("User-Agent", "Mozilla/5.0");
                    vh.put("Cookie", cookies);
                    vh.put("Referer", BASE_URL + "/");
                    HttpResult vr = HttpClient.get(BASE_URL + "/view/" + mailId, vh);
                    if (vr.isOk()) htmlBody = vr.getBody();
                } catch (RuntimeException ignored) {
                }
            }

            Map<String, Object> row = new LinkedHashMap<>();
            row.put("id", mailId);
            row.put("from", Json.str(msg, "from"));
            row.put("subject", Json.str(msg, "subject"));
            row.put("html", htmlBody);
            row.put("to", email);
            row.put("date", Json.str(msg, "date"));
            out.add(Normalizer.normalizeEmail(row, email));
        }
        return out;
    }
}

File: sdk/kotlin/src/main/kotlin/com/xxxxteam/tempmail/providers/Altmails.kt
package com.xxxxteam.tempmail.providers

import com.xxxxteam.tempmail.Email
import com.xxxxteam.tempmail.EmailInfo
import com.xxxxteam.tempmail.Http
import com.xxxxteam.tempmail.Normalize
import com.xxxxteam.tempmail.Provider
import kotlinx.serialization.json.*

/**
 * Altmails 渠道（tempmail.altmails.com）。
 *
 * GET / 取 session 和 CSRF，GET /random-email-address 取随机邮箱，
 * POST /fetch-emails/{email} 收信，GET /view/{id} 取正文。token 存储 JSON {csrf, cookies}。
 */
object Altmails : Provider {

    private const val BASE_URL = "https://tempmail.altmails.com"
    private val CSRF_RE = Regex("['\"]_token['\"]\\s*:\\s*['\"]([^'\"]+)['\"]")

    /** 创建临时邮箱。 */
    override suspend fun generate(): EmailInfo {
        val r1 = ProviderUtil.httpGet("$BASE_URL/",
            mapOf("Accept" to "text/html,application/xhtml+xml", "User-Agent" to "Mozilla/5.0"))
        r1.ensureSuccess()
        var cookies = cookieStr(r1)
        val csrf = CSRF_RE.find(r1.body)?.groupValues?.get(1)
            ?: throw RuntimeException("altmails: 无法从首页提取 CSRF token")
        val h2 = mapOf(
            "Accept" to "text/html", "User-Agent" to "Mozilla/5.0",
            "Cookie" to cookies, "Referer" to "$BASE_URL/",
        )
        val r2 = ProviderUtil.httpGet("$BASE_URL/random-email-address", h2)
        r2.ensureSuccess()
        cookies = mergeCookieStr(cookies, r2)
        val mailbox = r2.body.trim()
        if (mailbox.isBlank() || !mailbox.contains("@")) {
            throw RuntimeException("altmails: 返回的邮箱地址无效")
        }
        val token = buildJsonObject { put("csrf", csrf); put("cookies", cookies) }.toString()
        return EmailInfo(email = mailbox, channel = "altmails", token = token)
    }

    /** 获取邮件列表。 */
    override suspend fun getEmails(info: EmailInfo): List<Email> {
        val email = info.email.trim()
        if (email.isBlank()) throw RuntimeException("altmails: 邮箱地址为空")
        val tokObj = Http.json.parseToJsonElement(info.token) as? JsonObject
            ?: throw RuntimeException("altmails: token 格式无效")
        val csrf = jsonStr(tokObj, "csrf")
        val cookies = jsonStr(tokObj, "cookies")
        val h = mapOf(
            "Accept" to "application/json",
            "X-Requested-With" to "XMLHttpRequest",
            "Content-Type" to "application/x-www-form-urlencoded",
            "Cookie" to cookies, "Referer" to "$BASE_URL/", "User-Agent" to "Mozilla/5.0",
        )
        val resp = ProviderUtil.httpPost("$BASE_URL/fetch-emails/$email",
            "_token=${ProviderUtil.urlEncode(csrf)}", "application/x-www-form-urlencoded", h)
        resp.ensureSuccess()
        val arr = Http.json.parseToJsonElement(resp.body) as? JsonArray ?: return emptyList()
        return arr.mapNotNull { it as? JsonObject }.map { msg ->
            val mailId = jsonStr(msg, "id")
            val htmlBody = if (mailId.isNotEmpty()) fetchView(mailId, cookies) else ""
            val flat = mapOf(
                "id" to mailId,
                "from" to jsonStr(msg, "from"),
                "subject" to jsonStr(msg, "subject"),
                "html" to htmlBody,
                "to" to email,
                "date" to jsonStr(msg, "date"),
            )
            Normalize.fromMap(flat, email)
        }
    }

    /** 拉取单封邮件 HTML 正文，失败返回空串。 */
    private suspend fun fetchView(mailId: String, cookies: String): String {
        return try {
            val h = mapOf("User-Agent" to "Mozilla/5.0", "Cookie" to cookies, "Referer" to "$BASE_URL/")
            val vr = ProviderUtil.httpGet("$BASE_URL/view/$mailId", h)
            if (vr.isOk) vr.body else ""
        } catch (_: Exception) {
            ""
        }
    }

    /** 从响应 Set-Cookie 拼接 name=value; 串。 */
    private fun cookieStr(resp: HttpResp): String =
        resp.setCookies.joinToString("; ") { it.split(";")[0].trim() }

    /** 合并已有 cookie 串与响应新 cookie。 */
    private fun mergeCookieStr(prev: String, resp: HttpResp): String {
        val jar = LinkedHashMap<String, String>()
        prev.split(";").forEach { part ->
            val p = part.trim()
            val eq = p.indexOf('=')
            if (eq > 0) jar[p.substring(0, eq).trim()] = p.substring(eq + 1).trim()
        }
        resp.setCookies.forEach { sc ->
            val seg = sc.split(";")[0].trim()
            val eq = seg.indexOf('=')
            if (eq > 0) jar[seg.substring(0, eq)] = seg.substring(eq + 1)
        }
        return jar.entries.joinToString("; ") { "${it.key}=${it.value}" }
    }
}

/** 从 JSON 对象读取字符串字段，缺失返回空串（文件私有）。 */
private fun jsonStr(obj: JsonObject, key: String): String =
    (obj[key] as? JsonPrimitive)?.contentOrNull ?: ""
